Red thread (#89)
chhoumann/p7: report/sections/Chapter 02 - Preliminaries/preliminaries.tex

> The following chapter contains short introductions to the most important technologies used in the project. Additionally, we also introduce relevant theory.

Reviewer: ```suggestion
The following chapter contains short introductions to the most important technologies used in the project. Additionally, theory necessary to understand the various components of the developed platform is presented.
```

Rust is a multi-paradigm, general-purpose programming language originally developed by Graydon Hoare in 2006 as a personal project at Mozilla. 
The project was then sponsored by Mozilla from 2009 to 2021 where the project along with all trademarks was moved to the newly create Rust Foundation. 
The idea behind Rust was to challenge the idea that high-level ergonomics and low-level control come at the cost of each other\cite{Rust_Book}.

> Most other programming languages today operate in a band where either safety or control is valued highest and the other diminished. An example of this is C that has high level control but lacks safety.

Reviewer: ```suggestion
Most other programming languages today operate in a band where either safety or control is valued highest and the other diminished. An example of this is the C Language that offers high level of control but lacks safety. \todo{source}
```

Here Rust provides the low-level power of C along with the solid safety features programmers have gotten used to in high-level languages like Python. 
As such, the main overall goal of Rust is to provide as much safety as possible in a low-level programming language without sacrificing any speed\cite{Rust_in_Action}.

In order to achieve this goal Rust provides a large number of safety features, among these is its most distinguishing feature which prevents invalid data access at compile time. 
Considering a research article by Microsoft Security Research Center stating that invalid data access is behind 70\% of serious security bugs in modern programs, the benefits that Rust provide are significant\cite{Safe_Systems_Languages}. 

> Instead of relying on the developer to manage memory, Rust's compiler guarantees that any compiled program is memory-safe.

Reviewer: I would like to have a souce on guarantee of memory-safety.

\section*{Docker}
The main purpose of Docker is to simplify deliveries in distributed systems with the use of so-called containers\cite{Docker_Container}.
Containers are bundles of code and required dependencies. This allows developers to move around prepackaged applications ensuring that the developed software runs the same regardless of what system is used\cite{Docker_Container}.
The containers themselves are created from an image. Here the image is the executable package that encapsulates everything required in order to run the application. On runtime, the image is then turned into a container. 

> This is possible due to the Linux kernel's build process isolation and virtualization capabilities. These also allow a single host system to share its resources with multiple application components\cite{Docker_Container}.

Reviewer: "These also allow a single"   er "these" containers eller capabilities?

> We use Docker to set up local development environments for the APIs and Databases. This allows for an easier development process and a simplified delivery process to production.

Reviewer: Do we dockerize all production?

\section*{tRPC}
tRPC is a library that enables developers to build type safe APIs without having to deal with schemas or code generation. 
It is structured in a way that allows for types to be shared between the client and server by only importing the types instead of the actual server code. 
This in turn allows developers to not have the actual server code exposed in the \frontend{}. 

> Furthermore, tRPC is created in a way to fully take advantage of the power of a full-stack TypeScript project, which is a language-agnostic framework\cite{tRPC}.

Reviewer: How does it take advantage? Does "which" refer to tRPC or TS?

\section*{TypeScript}
As the server-side use of \javascript{} keeps growing in popularity its inability to provide developers with access to strong type checking and compile time error checking continues to prevent it from being the go-to language.
As a way to address these problems, \typescript{} was designed to try and bridge this gap without requiring developers to learn a new language.

> Instead, \typescript{} was developed as a syntactic superset to \javascript{} allowing developers to keep building into their already existing knowledge but with additional features. 

Reviewer: ```suggestion
Instead, \typescript{} was developed as a syntactic superset to \javascript{} allowing developers to extend existing knowledge and features with typesafe functionality.
```

> This also allows for all \javascript{} projects to be converted to \typescript{} projects allowing for easier movement between the two languages\cite{TypeScript}. 

Reviewer: ```suggestion
This means that all \javascript{} projects are syntactically valid in typescript. This in turn allows a development team to gradually adopt typescript into their project and use \typescript{} and \javascript{} interchangeably. \cite{TypeScript}. 
```

\typescript{} accomplishes this while still keeping the advantages of an interpreted language by compiling all the written instructions to its \javascript{} equivalents during the compilation process. 
This is possible because \typescript{} itself is compiled into \javascript{}, which is an interpreted language. 

\section*{Next.js}

> Next.js (Next) is an open-source React framework created by Vercel that enables react web applications to use server-side rendering as well as generate static websites. The Next framework was created in order to counter the potential security issues of React, which relies on client-side rendering. Next provides a solution to this by enabling websites to be fully or partially server-side rendered\cite {Nextjs_Docks}. 

Reviewer: ```suggestion
Next.js (Next) is an open-source React framework created by Vercel that enables react web applications to use server-side rendering as well as generate static websites. Next was created to counter the potential security issues of React, which relies on client-side rendering. Next provides a solution  by enabling websites to be fully or partially server-side rendered\cite {Nextjs_Docks}. 
```

Reviewer 2: We should probably refer to it as Next.js everywhere, as that's how the docs also refer to it.

> Next as a framework also tries to fix another issue of React, namely that creating a website from the ground up still requires a lot of effort, such as spending time on configuring tools as well as rebuilding solutions. 

Reviewer: ```suggestion
Next  tries to fix another issue of React, namely that creating a website from the ground up still requires a lot of effort, such as spending time on configuring tools as well as rebuilding solutions. 
```

Reviewer 2: We should probably refer to it as Next.js everywhere, as that's how the docs also refer to it.

BTW, double space between Next and tries.
